# Introduction
This notebook has two purposes:
1. Determine whether any duplicate samples have different collection years
2. Create summary of numbers of samples and variants for meeting with Dominic 01/02/2017

In [1]:
%run _standard_imports.ipynb

python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:53:06) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]
numpy 1.11.2
scipy 0.18.1
pandas 0.19.0
numexpr 2.6.1
pysam 0.8.4
petl 1.1.0
petlx 1.0.3
vcf 0.6.8
vcfnp 2.2.0
h5py 2.6.0
tables 3.3.0


In [26]:
panoptes_final_metadata_fn = "/nfs/team112_internal/rp7/data/methods-dev/builds/Pf6.0/20161214_samples_which_release/pf_6_0_panoptes_samples_final_20170124.txt.gz"
hdf_fn = "/nfs/team112_internal/production/release_build/Pf/6_0_release_packages/hdf5/Pf_60.h5"

In [50]:
tbl_panoptes_samples_final = (
    etl
    .fromtsv(panoptes_final_metadata_fn)
    .convertnumbers()
)
print(len(tbl_panoptes_samples_final.data()))
print(len(tbl_panoptes_samples_final.distinct('Sample').data()))
tbl_panoptes_samples_final

7182
7182


+------------+---------------+-----------------+---------------+---------------------+-------+--------+--------+------------------+------------------+------------+-------------+-------------+-----------+------+------+--------------+--------------------+-------------------------------------------+---------------------+-------------------------+-------------+------------------------+-----------------+
| Sample     | OxfordSrcCode | OxfordDonorCode | Individual_ID | AlfrescoStudyCode   | Type  | InV5.1 | InV5   | SubContinentCode | SubContinentName | RegionCode | CountryCode | CountryName | Site      | Year | Date | HasDuplicate | DiscardAsDuplicate | run_accessions                            | pc_pass_missing     | pc_genome_covered_at_1x | Year_source | sort_AlfrescoStudyCode | IndividualGroup |
+============+===============+=================+===============+=====================+=======+========+========+==================+==================+============+=============+=============+===========+======+======+==============+====================+===========================================+=====================+=========================+=============+========================+=================+
| 'PM0001-C' | '215C'        | '215C'          | '215C'        | '1001-PF-ML-DJIMDE' | 'DNA' | 'True' | 'True' | 'WAF'            | 'West Africa'    | 'ML'       | 'ML'        | 'Mali'      | 'Kolle'   | 2007 | 2007 | 'False'      | 'False'            | 'ERR012350,ERR012360,ERR012305'           | 0.15139786050931692 |                  0.9696 | 'study'     | '1001-PF-ML-DJIMDE'    |               1 |
+------------+---------------+-----------------+---------------+---------------------+-------+--------+--------+------------------+------------------+------------+-------------+-------------+-----------+------+------+--------------+--------------------+-------------------------------------------+---------------------+-------------------------+-------------+------------------------+-----------------+
| 'PM0002-C' | '217C'        | '217C'          | '217C'        | '1001-PF-ML-DJIMDE' | 'DNA' | 'True' | 'True' | 'WAF'            | 'West Africa'    | 'ML'       | 'ML'        | 'Mali'      | 'Kolle'   | 2007 | 2007 | 'False'      | 'False'            | 'ERR012520,ERR012331,ERR012327'           | 0.28678549872221293 |                   0.864 | 'study'     | '1001-PF-ML-DJIMDE'    |               2 |
+------------+---------------+-----------------+---------------+---------------------+-------+--------+--------+------------------+------------------+------------+-------------+-------------+-----------+------+------+--------------+--------------------+-------------------------------------------+---------------------+-------------------------+-------------+------------------------+-----------------+
| 'PM0003-C' | '520C'        | '520C'          | '520C'        | '1001-PF-ML-DJIMDE' | 'DNA' | 'True' | 'True' | 'WAF'            | 'West Africa'    | 'ML'       | 'ML'        | 'Mali'      | 'Kolle'   | 2007 | 2007 | 'False'      | 'False'            | 'ERR012492,ERR012457,ERR012532'           |  0.1662372060768727 |                  0.9586 | 'study'     | '1001-PF-ML-DJIMDE'    |               3 |
+------------+---------------+-----------------+---------------+---------------------+-------+--------+--------+------------------+------------------+------------+-------------+-------------+-----------+------+------+--------------+--------------------+-------------------------------------------+---------------------+-------------------------+-------------+------------------------+-----------------+
| 'PM0004-C' | 'A336TPI7C'   | 'A336TPI7C'     | 'A336TPI7C'   | '1001-PF-ML-DJIMDE' | 'DNA' | 'True' | 'True' | 'WAF'            | 'West Africa'    | 'ML'       | 'ML'        | 'Mali'      | 'Kolle'   | 2007 | 2007 | 'False'      | 'False'            | 'ERR015438,ERR015329,ERR015422,ERR015315' | 0.11811536041300132 |                  0.9648 | 'study'     | '1001-PF-ML-D

In [51]:
(
    tbl_panoptes_samples_final
    .selecteq('HasDuplicate', 'True')
    .valuecounts('IndividualGroup', 'Site', 'Year')
    .selecteq('count', 1)
    .sort(('IndividualGroup', 'Site', 'Year'))
    .displayall()
)

IndividualGroup 
 Site 
 Year 
 count 
 frequency 
 
 
 
 
 1939 
 Pursat 
 2013 
 1 
 0.0012135922330097086 
 
 
 1940 
 Pursat 
 2013 
 1 
 0.0012135922330097086 
 
 
 1962 
 Pursat 
 2014 
 1 
 0.0012135922330097086 
 
 
 1962 
 Pursat 
 2015 
 1 
 0.0012135922330097086 
 
 
 2060 
 Preah Vihear 
 2013 
 1 
 0.0012135922330097086 
 
 
 2067 
 Preah Vihear 
 2013 
 1 
 0.0012135922330097086 
 
 
 2160 
 Ratanakiri 
 2013 
 1 
 0.0012135922330097086 
 
 
 2160 
 Ratanakiri 
 2014 
 1 
 0.0012135922330097086 
 
 
 6614 
 Timika 
 2012 
 1 
 0.0012135922330097086 
 
 
 6614 
 Timika 
 2011† 
 1 
 0.0012135922330097086 
 
 
 6687 
 Timika 
 2011 
 1 
 0.0012135922330097086 
 
 
 6687 
 Timika 
 2011† 
 1 
 0.0012135922330097086

In [52]:
problematic_groups = (
    tbl_panoptes_samples_final
    .selecteq('HasDuplicate', 'True')
    .valuecounts('IndividualGroup', 'Site', 'Year')
    .selecteq('count', 1)
    .sort(('IndividualGroup', 'Site', 'Year'))
    .distinct('IndividualGroup')
    .values('IndividualGroup')
    .array()
)
problematic_groups

array([1939, 1940, 1962, 2060, 2067, 2160, 6614, 6687])

In [53]:
(
    tbl_panoptes_samples_final
    .selectin('IndividualGroup', problematic_groups)
    .cut(['Sample', 'OxfordSrcCode', 'Individual_ID', 'AlfrescoStudyCode', 'Site', 'Year', 'DiscardAsDuplicate', 'IndividualGroup'])
    .displayall()
)

Sample 
 OxfordSrcCode 
 Individual_ID 
 AlfrescoStudyCode 
 Site 
 Year 
 DiscardAsDuplicate 
 IndividualGroup 
 
 
 
 
 PH1025-C 
 163-KH1-032 
 163-KH1-032 
 1044-PF-KH-FAIRHURST 
 Pursat 
 2012 
 False 
 1939 
 
 
 PH1090-C 
 163-KH1-032RME 
 163-KH1-032 
 1044-PF-KH-FAIRHURST 
 Pursat 
 2013 
 True 
 1939 
 
 
 PH1347-C 
 163-KH1-032-CF11-adap 
 163-KH1-032 
 1044-PF-KH-FAIRHURST 
 Pursat 
 2012 
 True 
 1939 
 
 
 PH1381-C 
 163-KH1-032-adap 
 163-KH1-032 
 1044-PF-KH-FAIRHURST 
 Pursat 
 2012 
 True 
 1939 
 
 
 PH1026-C 
 163-KH1-033 
 163-KH1-033 
 1044-PF-KH-FAIRHURST 
 Pursat 
 2012 
 True 
 1940 
 
 
 PH1091-C 
 163-KH1-033RME 
 163-KH1-033 
 1044-PF-KH-FAIRHURST 
 Pursat 
 2013 
 True 
 1940 
 
 
 PH1329-C 
 163-KH1-033-CF11-adap 
 163-KH1-033 
 1044-PF-KH-FAIRHURST 
 Pursat 
 2012 
 True 
 1940 
 
 
 PH1336-C 
 163-KH1-033RME-CF11-adap 
 163-KH1-033 
 1044-PF-KH-FAIRHURST 
 Pursat 
 2012 
 False 
 1940 
 
 
 PH1362-C 
 163-KH1-033-adap 
 163-KH1-033 
 1044-PF-KH-FAIRHURST 
 Pursat 
 2012 
 True 
 1940 
 
 
 PH1369-C 
 163-KH1-033RME-adap 
 163-KH1-033 
 1044-PF-KH-FAIRHURST 
 Pursat 
 2012 
 True 
 1940 
 
 
 PH1205-C 
 163-KH1-062 
 163-KH1-062 
 1044-PF-KH-FAIRHURST 
 Pursat 
 2013 
 False 
 1962 
 
 
 PH1266-C 
 163-KH1-062RME 
 163-KH1-062 
 1044-PF-KH-FAIRHURST 
 Pursat 
 2013 
 True 
 1962 
 
 
 PH1400-C 
 163-KH1-062-2-Filter 
 163-KH1-062 
 1044-PF-KH-FAIRHURST 
 Pursat 
 2014 
 True 
 1962 
 
 
 PH1400-CW2 
 163-KH1-062-2 
 163-KH1-062 
 1044-PF-KH-FAIRHURST 
 Pursat 
 2015 
 True 
 1962 
 
 
 PH1039-C 
 163-KH2-012 
 163-KH2-012 
 1044-PF-KH-FAIRHURST 
 Preah Vihear 
 2012 
 True 
 2060 
 
 
 PH1098-C 
 163-KH2-012RME 
 163-KH2-012 
 1044-PF-KH-FAIRHURST 
 Preah Vihear 
 2013 
 True 
 2060 
 
 
 PH1348-C 
 163-KH2-012RME-CF11-adap 
 163-KH2-012 
 1044-PF-KH-FAIRHURST 
 Preah Vihear 
 2012 
 True 
 2060 
 
 
 PH1355-C 
 163-KH2-012-CF11-adap 
 163-KH2-012 
 1044-PF-KH-FAIRHURST 
 Preah Vihear 
 2012 
 False 
 2060 
 
 
 PH1382-C 
 163-KH2-012RME-adap 
 163-KH2-012 
 1044-PF-KH-FAIRHURST 
 Preah Vihear 
 2012 
 True 
 2060 
 
 
 PH1389-C 
 163-KH2-012-adap 
 163-KH2-012 
 1044-PF-KH-FAIRHURST 
 Preah Vihear 
 2012 
 True 
 2060 
 
 
 PH1099-C 
 163-KH2-021RME 
 163-KH2-021 
 1044-PF-KH-FAIRHURST 
 Preah Vihear 
 2013 
 True 
 2067 
 
 
 PH1349-C 
 163-KH2-021RME-CF11-adap 
 163-KH2-021 
 1044-PF-KH-FAIRHURST 
 Preah Vihear 
 2012 
 True 
 2067 
 
 
 PH1353-C 
 163-KH2-021-CF11-adap 
 163-KH2-021 
 1044-PF-KH-FAIRHURST 
 Preah Vihear 
 2012 
 True 
 2067 
 
 
 PH1383-C 
 163-KH2-021RME-adap 
 163-KH2-021 
 1044-PF-KH-FAIRHURST 
 Preah Vihear 
 2012 
 True 
 2067 
 
 
 PH1387-C 
 163-KH2-021-adap 
 163-KH2-021 
 1044-PF-KH-FAIRHURST 
 Preah Vihear 
 2012 
 False 
 2067 
 
 
 PH1316-C 
 163-KH3-062 
 163-KH3-062 
 1044-PF-KH-FAIRHURST 
 Ratanakiri 
 2013 
 False 
 2160 
 
 
 PH1322-C 
 163-KH3-062RME 
 163-KH3-062 
 1044-PF-KH-FAIRHURST 
 Ratanakiri 
 2014 
 True 
 2160 
 
 
 PJ0161-C 
 UNKNOWN_UFC_104 
 UFC_104 
 1155-PF-ID-PRICE 
 Timika 
 2011† 
 True 
 6614 
 
 
 PJ0183-C 
 UFC_104 
 UFC_104 
 1155-PF-ID-PRICE 
 Timika 
 2012 
 False 
 6614 
 
 
 PJ0134-C 
 UNKNOWN_UFM_32 
 UFM_32 
 1155-PF-ID-PRICE 
 Timika 
 2011† 
 False 
 6687 
 
 
 PJ0253-C 
 UFM_32 
 UFM_32 
 1155-PF-ID-PRICE 
 Timika 
 2011 
 True 
 6687

# Duplicate samples conclusion
There are a few samples from study 1044 that look like probable recrudescences. Two sample pairs from 1155 with different dates are due to assigned samples to 2011† in elife paper. Given the samples from 1044, we should probably not release collection year for discarded samples.

# Sample numbers

In [54]:
# Are there any samples with duplicates one of which was in v5 and the other of which is new?
v5_v6_groups = (
    tbl_panoptes_samples_final
    .selecteq('HasDuplicate', 'True')
    .valuecounts('IndividualGroup', 'InV5')
    .selecteq('count', 1)
    .sort(('IndividualGroup', 'InV5'))
    .distinct('IndividualGroup')
    .values('IndividualGroup')
    .array()
)
v5_v6_groups

array([ 650, 1353, 6614, 6683, 6687])

In [63]:
(
    tbl_panoptes_samples_final
    .selectin('IndividualGroup', v5_v6_groups)
    .cut(['Sample', 'OxfordSrcCode', 'Individual_ID', 'AlfrescoStudyCode', 'InV5', 'DiscardAsDuplicate', 'pc_pass_missing', 'IndividualGroup'])
    .displayall()
)

Sample 
 OxfordSrcCode 
 Individual_ID 
 AlfrescoStudyCode 
 InV5 
 DiscardAsDuplicate 
 pc_pass_missing 
 IndividualGroup 
 
 
 
 
 PH0027-C 
 CP268 
 CP268 
 1011-PF-KH-SU 
 True 
 True 
 0.34629377544337286 
 650 
 
 
 PH0027-Cx 
 CP268 
 CP268 
 1011-PF-KH-SU 
 False 
 False 
 0.0699845252924784 
 650 
 
 
 PT0229-C 
 H222 
 H222 
 1022-PF-MW-OCHOLLA 
 True 
 False 
 0.006057288523032272 
 1353 
 
 
 PT0229-Cx 
 H222 
 H222 
 1022-PF-MW-OCHOLLA 
 False 
 True 
 0.9979873248661213 
 1353 
 
 
 PJ0161-C 
 UNKNOWN_UFC_104 
 UFC_104 
 1155-PF-ID-PRICE 
 True 
 True 
 0.06205518242175962 
 6614 
 
 
 PJ0183-C 
 UFC_104 
 UFC_104 
 1155-PF-ID-PRICE 
 False 
 False 
 0.0387805159948118 
 6614 
 
 
 PJ0129-CW 
 UFM_26 
 UFM_26 
 1155-PF-ID-PRICE 
 False 
 True 
 0.042826092540035186 
 6683 
 
 
 PJ0129-Cx 
 UFM_26 
 UFM_26 
 1155-PF-ID-PRICE 
 True 
 False 
 0.03205287084719208 
 6683 
 
 
 PJ0134-C 
 UNKNOWN_UFM_32 
 UFM_32 
 1155-PF-ID-PRICE 
 True 
 False 
 0.013303432688232801 
 6687 
 
 
 PJ0253-C 
 UFM_32 
 UFM_32 
 1155-PF-ID-PRICE 
 False 
 True 
 0.025698930254658466 
 6687

In [56]:
tbl_panoptes_samples_final.pivot('InV5', 'DiscardAsDuplicate', 'DiscardAsDuplicate', len)

+---------+-------+------+
| InV5    | False | True |
+=========+=======+======+
| 'False' |  1030 |  159 |
+---------+-------+------+
| 'True'  |  5666 |  327 |
+---------+-------+------+

In [57]:
tbl_panoptes_samples_final.selectne('Site', 'Lab').pivot('InV5', 'DiscardAsDuplicate', 'DiscardAsDuplicate', len)

+---------+-------+------+
| InV5    | False | True |
+=========+=======+======+
| 'False' |   963 |  159 |
+---------+-------+------+
| 'True'  |  5658 |  326 |
+---------+-------+------+

In [58]:
tbl_panoptes_samples_final.selectne('Year', 'Lab').pivot('InV5', 'DiscardAsDuplicate', 'DiscardAsDuplicate', len)

+---------+-------+------+
| InV5    | False | True |
+=========+=======+======+
| 'False' |   963 |  159 |
+---------+-------+------+
| 'True'  |  5658 |  326 |
+---------+-------+------+

In [61]:
tbl_panoptes_samples_final.selectge('pc_genome_covered_at_1x', 0.5).selectne('Year', 'Lab').pivot('InV5', 'DiscardAsDuplicate', 'DiscardAsDuplicate', len)

+---------+-------+------+
| InV5    | False | True |
+=========+=======+======+
| 'False' |   893 |  150 |
+---------+-------+------+
| 'True'  |  5551 |  318 |
+---------+-------+------+

In [48]:
print(len(tbl_panoptes_samples_final.data()))
print(len(tbl_panoptes_samples_final.selectne('Year', 'Lab').data()))
print(len(tbl_panoptes_samples_final.selecteq('Year', 'Lab').data()))


7182
7106
76


# Variant numbers

In [27]:
callset = h5py.File(hdf_fn, mode='r')
variants = callset['variants']
PASS = callset['variants']['FILTER_PASS'][:]
is_snp = callset['variants']['is_snp'][:]

In [34]:
num_variants = collections.OrderedDict()
num_variants['Pass SNPs'] = np.count_nonzero(PASS & is_snp)
num_variants['Fail SNPs'] = np.count_nonzero(np.logical_not(PASS) & is_snp)
num_variants['Pass indels'] = np.count_nonzero(PASS & np.logical_not(is_snp))
num_variants['Fail indels'] = np.count_nonzero(np.logical_not(PASS) & np.logical_not(is_snp))
print(len(PASS))
print(len(is_snp))
num_variants

6051696
6051696


OrderedDict([('Pass SNPs', 1838733),
             ('Fail SNPs', 1329988),
             ('Pass indels', 1276027),
             ('Fail indels', 1606948)])

In [35]:
1838733+1329988+1276027+1606948

6051696

# Sample metadata

In [36]:
tbl_panoptes_samples_final.header()

('Sample',
 'OxfordSrcCode',
 'OxfordDonorCode',
 'Individual_ID',
 'AlfrescoStudyCode',
 'Type',
 'InV5.1',
 'InV5',
 'SubContinentCode',
 'SubContinentName',
 'RegionCode',
 'CountryCode',
 'CountryName',
 'Site',
 'Year',
 'Date',
 'HasDuplicate',
 'DiscardAsDuplicate',
 'run_accessions',
 'pc_pass_missing',
 'pc_genome_covered_at_1x',
 'Year_source',
 'sort_AlfrescoStudyCode',
 'IndividualGroup')

In [37]:
variables_to_count = [
 'AlfrescoStudyCode',
 'SubContinentCode',
 'SubContinentName',
 'RegionCode',
 'CountryCode',
 'CountryName',
 'Site',
 'Year']

In [44]:
for variable in variables_to_count:
    print(variable, len(tbl_panoptes_samples_final.selectne('Site', 'Lab').distinct(variable).data()))

AlfrescoStudyCode 52
SubContinentCode 9
SubContinentName 9
RegionCode 35
CountryCode 29
CountryName 29
Site 78
Year 16


In [47]:
tbl_panoptes_samples_final.selectne('Site', 'Lab').distinct('SubContinentName').cut('SubContinentName').displayall()

SubContinentName 
 
 
 
 
 Central Africa 
 
 
 East Africa 
 
 
 Eastern south-east Asia 
 
 
 Insular Africa 
 
 
 Oceania 
 
 
 South America 
 
 
 South Asia 
 
 
 West Africa 
 
 
 Western south-east Asia